File: 2178.py
# 미로 탐색
# https://www.acmicpc.net/problem/2178

import sys

input = sys.stdin.readline


def is_valid_index(graph, r, c):
    if r < 0 or c < 0:
        return False

    if len(graph) <= r:
        return False

    if len(graph[r]) <= c:
        return False

    return True


n, m = map(int, input().split())
graph = []

for _ in range(n):
    graph.append(list(map(int, list(input().strip()))))

adjacent_indices = [(1, 0), (-1, 0), (0, 1), (0, -1)]
queue = [(0, 0, 1)]
graph[0][0] = 0
answer = -1
break_flag = False
while len(queue) > 0 and not break_flag:
    cur_r, cur_c, count = queue.pop(0)

    for adjacent_index in adjacent_indices:
        next_r = cur_r + adjacent_index[0]
        next_c = cur_c + adjacent_index[1]

        if is_valid_index(graph, next_r, next_c):
            if graph[next_r][next_c] == 1:

                if next_r == n - 1 and next_c == m - 1:
                    answer = count + 1
                    break_flag = True
                    break

                graph[next_r][next_c] = 0
                queue.append((next_r, next_c, count + 1))

print(answer)

File: 2178.kt
// https://www.acmicpc.net/problem/2178
// 미로 탐색

import java.util.*

val maze = mutableListOf<List<Int>>()
val distance = mutableListOf<MutableList<Int>>()

val adjacent = listOf(0 to 1, 1 to 0, -1 to 0, 0 to -1)

fun main() {
    input()
    solve()
    val distR = maze.size - 1
    val distC = maze[distR].size - 1
    println(distance[distR][distC])
}

fun input() {
    val reader = System.`in`.bufferedReader()
    val (n, m) = reader.readLine().split(" ").map { it.toInt() }
    for (i in 0 until n) {
        val tmpRow = reader.readLine().map { it.digitToInt() }
        maze.add(tmpRow)
        distance.add(MutableList(m) { -1 })
    }
}

fun solve() {
    val queue = LinkedList<Pair<Int, Int>>()
    queue.add(0 to 0)
    distance[0][0] = 1
    while (queue.isNotEmpty()) {
        val current = queue.poll()
        val r = current.first
        val c = current.second
        val nextDistance = distance[r][c] + 1

        for (index in adjacent) {
            val nextR = r + index.first
            val nextC = c + index.second
            if (!isValidIndex(nextR, nextC)) continue

            if (maze[nextR][nextC] != 1) continue
            if (distance[nextR][nextC] != -1) continue

            distance[nextR][nextC] = nextDistance
            queue.add(nextR to nextC)
        }
    }
}

fun isValidIndex(r: Int, c: Int): Boolean {
    if (r < 0 || c < 0) return false
    if (maze.size <= r) return false
    if (maze[r].size <= c) return false
    return true
}